---
title: "MyTools"
author: "whzemuch"
date: "June 4, 2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## How to do 4-parametric regression for ELISA data in R
  - https://stats.stackexchange.com/questions/61144/how-to-do-4-parametric-regression-for-elisa-data-in-r
  
```{r eval=FALSE, include=FALSE}
install.packages("drc")
install.packages("curl")

library(drc)
library(dplyr)





```


## Calibration of Periotron 8000 and 6000 by polynomial regression
   J Periodontal Res. 1999 Feb;34(2):79-86.
    Clalibration of the Periotron 8000 and 6000 by polynomial regression.
    Chapple IL1, Landini G, Griffiths GS, Patel NC, Ward RS.
    PMID: 10207836
 - https://stats.stackexchange.com/questions/95939/how-to-interpret-coefficients-from-a-polynomial-model-fit   
    
```{r}

x <- c(32,64,96,118,126,144,152.5,158)  
y <- c(99.5,104.8,108.5,100,86,64,35.3,15)

modI <- lm(y ~ x + I(x^2) + I(x^3))
modP <- lm(y ~ poly(x,3, raw=T))

# predict y using modI
coef(modI)[1] + coef(modI)[2]*x + coef(modI)[3]*x^2 + coef(modI)[4]*x^3

# predict y using modP

x_ploy <-  stats:::predict.poly(object = poly(x, 3), newdata = 3)
coef(modP)[1] + coef(modP)[2]*x_ploy[1] + coef(modP)[3]*x_ploy[2] + coef(modP)[4]*x_ploy[3]


```

    
    
    


## Make 8x12 table for ELISA
  - id: Sample id
  - time: time points
  - std: standards
  - Output is a list that contain a matrix and a data frame
    1. matrix will be used for the layout of the experiment
    2. data frame will be  used for collecting data 

```{r}
qw <- function(...) {
  ## mimic qw funciton in Perl  
  ## https://stackoverflow.com/questions/520810/does-r-have-quote-like-operators-like-perls-qw
  sapply(match.call()[-1], deparse)
}

makeTable <- function(id, time, date, std) {
  # purpose: Make 8x12 table and the corresponding list for ELISA
  
 ids <- unlist(lapply(id, function(x) rep(x,length(time))))
 times <- rep(time, length(id))
 
 ll <- list(ids, times, date)
 df <-  data.frame(ratID=ids,
                   time=times,
                   date=date)
                   
 ll <-  do.call("paste", c(ll, sep = "; "))
 ll <- append(ll, rep("NA", 8-length(ll)%%8))
 llist <- matrix(ll, nrow=8)
 mat <- cbind(std, std, llist)
 colnames(mat) <- seq(1:length(colnames(mat)))
 rownames(mat) <- LETTERS[1:8]
 
 data <- list(mat, df, ll)
 
 
 return(data)
  
}

# idx <- c("N1", "N2", "N3", "N4", "45", "49", "53", "55", "57", "58", "59", "61")

# Prepare the parameter for makeTable function
id <- qw(N1, N2, N3, N4, 45, 48, 49, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64)
exp_date <- rep("5/14/19", 20)
time <- c("0 min", "15 min", "30 min", "60 min", "120 min")
std <- rev(2^seq(from=0, to=6, by=1)/10)
std <- c(std, 0)


data <-  makeTable(id, time, exp_date,  std)  # return all the needed objexts

  # write.csv(data[1], file = "./output/table.csv")
  # write.csv(data[2], file = "./output/dataframe.csv")
  # write.csv(data[3], file = "./output/list.csv")
  
  

# write data to csv file
  # the easly way: ------------------------------------------------
  # write.csv(data[1], file = "./output/table.csv")
  # write.csv(data[2], file = "./output/dataframe.csv")
  # write.csv(data[3], file = "./output/list.csv")

  # the better way described as below------------------------------
  # https://stat.ethz.ch/pipermail/r-help/2011-March/272076.html 

out_file <-  file("data4ELISA.csv", open="a") # create a file in append mode
path_file <- paste0(rstudioapi::getActiveDocumentContext()$path, "\n") # alternative getwd() can be used.
cat(path_file, "\n", file = out_file, append = TRUE) # write a message about the filename of the current script

for (i in seq_along(data)) {
  
  # write.table(names(data)[i], file = out_file, sep = ",", dec = ".", quote = FALSE,    col.names = FALSE, row.names = FALSE)  # write the name of the list elements
  
  write.table(data[i], file = out_file, sep = ",", dec = ".", quote = FALSE, col.names = FALSE, row.names = FALSE)  # write the data to csv file 
  
  cat("\n\n", file = out_file, append = TRUE)  # insert a blank line https://stat.ethz.ch/pipermail/r-help/2010-January/223660.html 
}

close(out_file)  # close connection to csv file







library(kableExtra)
# http://haozhu233.github.io/kableExtra/awesome_table_in_html_cn.html#%E8%A1%8C%E7%BB%84%E5%92%8C%E5%88%97%E7%BB%84

mat %>%
  kable() %>%
  kable_styling(bootstrap_options = c("striped", "hover")) %>% 
  save_kable(file = "table1.html", self_contained = T)








```

## Save lists to csv
  - ttps://stackoverflow.com/questions/19330949/r-how-to-save-lists-into-csv
   

```{r eval=FALSE, include=FALSE}

lapply(data, function(x) write.table(data.frame(x), "tetst.csv", append = T, sep = ","))
library(plyr) 
data.df = ldply(data, function(t) t$toDataFrame())
data.df <- do.call("rbind", lapply(data, as.data.frame))

```

 

## Hash in R

```{r}

# devtools::install_github("mkuhn/dict")
library(dict)
# https://www.rdocumentation.org/packages/dict/versions/0.10.0

d <- dict()

d[[c("45", "49")]] <-  "MSC"



```


## Create a multi-paneled plotting window.
layout, par(mfrow)
- http://rfunction.com/archives/1538

```{r}
 set.seed(5)
 x <- rnorm(200)
 y <- 25 - 22*x + 5*x^2 + rnorm(200)
 
 
 png("layout1.png")
 par(mfrow=c(2,2))
 par(mar=c(0.5, 4.5, 0.5, 0.5))
 boxplot(x, horizontal=TRUE, axes=FALSE)
 plot(0, type="n", xlab="", ylab="", axes=FALSE)
 par(mar=c(4.5, 4.5, 0.5, 0.5))
 plot(x, y)
 text(0.5, 85, "par(mfrow)", cex=2)
 par(mar=c(4.5, 0.5, 0.5, 0.5))
 boxplot(y, axes=FALSE)
 dev.off()
 

 
 png("layout2.png")
 mat <- matrix(c(1,2,0,3), 2)
     
 layout(mat, c(3.5,1), c(1,3))
 par(mar=c(0.5, 4.5, 0.5, 0.5))
 boxplot(x, horizontal=TRUE, axes=FALSE)
 par(mar=c(4.5, 4.5, 0.5, 0.5))
 plot(x, y)
 text(0.5, 85, "layout", cex=2)
 par(mar=c(4.5, 0.5, 0.5, 0.5))
 boxplot(y, axes=FALSE)
 dev.off()

```

## scatterplot  matrix
 -https://www.stat.auckland.ac.nz/~ihaka/787/lectures-layouts.pdf

```{r}
x <- iris[,1:4]


opar = par(mfrow = c(ncol(x), ncol(x)),
           mar = rep(0, 4),
           oma = rep(4.1, 4))

par(opar)


for(i in 1:4) {
  for(j in 1:4) {
    xrange = range(x[,j])
    yrange = range(x[,i])
    plot.new()
    plot.window(xlim = xrange, ylim = yrange)
    if (i == j)
      text(mean(xrange), mean(yrange),
      colnames(x)[j])
    else points(x[,j], x[,i])
    box()
  }
} 



i=1
j=4

    xrange = range(x[,i])
    yrange = range(x[,i])
    plot.new()
    plot.window(xlim = xrange, ylim = yrange)
    points(x[,1], x[,4])
    box()


```

```{r}
# http://www.sthda.com/english/wiki/scatter-plot-matrices-r-base-graphs
pairs(iris[,1:4], pch = 19)

pairs(iris[,1:4], pch = 19, lower.panel = NULL)

upper.panel<-function(x, y){
  points(x,y, pch=19, col=c("red", "green3", "blue")[iris$Species])
  r <- round(cor(x, y), digits=2)
  txt <- paste0("R = ", r)
  usr <- par("usr"); on.exit(par(usr))
  par(usr = c(0, 1, 0, 1))
  text(0.5, 0.9, txt)
}
pairs(iris[,1:4], lower.panel = NULL, 
      upper.panel = upper.panel)

```


```{r}
library(tidyverse)
ggplot(data = mpg) + 
  geom_point(mapping = aes(x = displ, y = hwy)) +
  facet_grid(. ~ cyl)

ggplot(data = mpg) + 
  geom_point(mapping = aes(x = displ, y = hwy)) +
  facet_grid(drv ~ .)



```

```{r pattern}
a <- "\\"
b <- "\\\\"

d <- "\\*\\'"
e <- "\\*\\'\\\\"


writeLines(a)    # \
writeLines(b)    # \\

writeLines(d)   # \*\'
writeLines(e)   # \*\'\\ 


# What patterns will the regular expression \..\..\.. match? How would you represent it as a string?

# Mathch .one character.one character. one character

x <- "\\..\\..\\.."
writeLines(x)   # \..\..\..

z <- "\""

writeLines(z) # "

y <- "\"\\$\\^\\$\""

writeLines(y)   # "\$\^\$"

xx <- "ab\"$^$\"cd"
writeLines(xx)  # ab"$^$"cd
str_view(xx, y)



```

## UpSetR or UpSetPlot
- UpSetR：集合可视化神包 by hoptop https://www.jianshu.com/p/324aae3d5ea4
- ChIPseeker的upsetplot是怎么写的 by hoptop https://www.jianshu.com/p/a01c426de160
- 从零开始学技能，以数据集合upset图为例。 by 果子 https://mp.weixin.qq.com/s/ybLITsqdErWCGF9hfVD2-g

```{r}
# ggupset:https://github.com/const-ae/ggupset

library(ggupset)
library(tidyverse)
library(data.table)
library(UpSetR)

glimpse(tidy_movies)

p <- tidy_movies %>%
  distinct(title, year, length, .keep_all=TRUE) %>%
  ggplot(aes(x=Genres)) +
    geom_bar() +
    scale_x_upset(n_intersections = 20)

x <- tidy_movies %>%
  distinct(title, year, length, .keep_all=TRUE)


# ggupsetr

# I copied the image for "gene_pathway_membership.rda"  and OCR to excel file using https://www.onlineocr.net/
file <- "./data/pathway_upsetR.csv" # I changed  last colomun last 2 rows from false to ture
data <- fread(file = file, header = T, sep = ",")
colnames(data)[1] <- "Pathway"

head(data)

tidy_data <- data %>%
  gather(Gene, Member, -Pathway) %>%
  filter(Member=="TRUE") %>%
  select(- Member) %>% 
  group_by(Gene) %>%
  summarize(Pathways = list(Pathway))

head(tidy_data)

p.ggupset <- ggplot(tidy_data,aes(x = Pathways)) +
  geom_bar() +
  scale_x_upset()

# UpSetR

# three methods for convert first column to rownames

data3 <- data.frame(data, row.names = 1)

data2 <- data %>% 
   remove_rownames %>% 
   column_to_rownames(var="Pathway") %>% 
   t() %>% 
   as.data.frame()

data2 <-  data2*1
data2 <-  data2[,-1]  # first column are all 0, which cause the error 
                      # Error in `[.data.frame`(data, , start_col:end_col) : 
                      #      undefined columns selected


head(data2)

p.upsetr <- upset(data2, sets = colnames(data2), keep.order = TRUE)


# ggupsetr with inset (venn diagram)

p1.ggupset <- ggplot(tidy_data,aes(x = Pathways)) +
  geom_bar() +
  theme(panel.border = element_blank(), 
         panel.grid.major = element_blank(),
         panel.grid.minor = element_blank(), 
         axis.line = element_line(colour = "black"))+
  scale_x_upset()

## venn diagram: yyplot::ggvenn; data: data2


# library(devtools)
# devtools::install_github('GuangchuangYu/yyplot', force=T)
library(yyplot)


p1.venn <- yyplot::ggvenn(data2)+
    theme_void()+
    theme(legend.position = "none")

library(ggimage)

p3 <- as.ggplot(p1.ggupset) + geom_subview(subview = p1.venn, x=.7, y=.8, w=.3, h=.3)

```
## 箱线图加上点图才是方向。
https://mp.weixin.qq.com/s/EC81-45FaFO-2oMDOt8vHg  Original:  果子

```{r}
# Knowledge1: new method for rerange factor levels
# Knowledge2: ggforce::geom_sina


df <-  data.frame(
  "Distribution" = c(rep("Unimodal", 500),
                     rep("Bimodal", 250), 
                     rep("Trimodal", 600)),
  "Value" = c(rnorm(500, 6, 1),
              rnorm(200, 3, .7), rnorm(50, 7, 0.4),
              rnorm(200, 2, .7), rnorm(300, 5.5,0.4), rnorm(100, 8, 0.4))
)

df$Distribution <- factor(df$Distribution, levels( df$Distribution)[c(2,3,1)])


library(ggplot2)
library(ggforce)

p <- ggplot(df, aes(Distribution, Value)) 

# barplot 
p + geom_col(aes(fill = Distribution))

# boxplot
p + geom_boxplot(aes(fill=Distribution))

# boxplot+ jitter
p + geom_boxplot(aes(fill=Distribution)) +
  geom_jitter()

# violin plot
p + geom_violin(aes(fill = Distribution))

# violin plot + jitter
p + geom_violin(aes(fill = Distribution)) +
  geom_jitter()

# vioiin plot + jitter(but keep dots inside)
p + geom_violin(aes(fill = Distribution)) +
    geom_sina()

# violin plot + dots
p + geom_sina(aes(color = Distribution))

```

## pdftools package
1. https://www.r-bloggers.com/pdftools-2-0-powerful-pdf-text-extraction-tools/
2. 自动化批量从PDF里面提取表格 https://mp.weixin.qq.com/s/5FJDyd7R7_RV6bWuTK0BRg
3. Extracting PDF Text with R and Creating Tidy Datahttps://blog.datazar.com/extracting-pdf-text-with-r-and-creating-tidy-data-f399011549cc

```{r}
library(pdftools)
library(tidyverse)


pdf_file <- "https://github.com/ropensci/tabulizer/raw/master/inst/examples/data.pdf"
txt <- pdf_text(pdf_file) 
txt1 <- readr::read_lines(txt[1])



# Extract first column (car model) using regular expression
txt_1col <- txt1[-1] %>% 
  str_extract_all("[:alpha:]+[a-zA-Z0-9 -]+\\s") %>% 
  unlist()
# Extract 2-last column
txt_2col_End <- txt1[-1] %>% 
  str_replace_all("[:alpha:]+[a-zA-Z0-9 -]+\\s", "")

# use cbind to combine them together, here read.table did great!
df <- data.frame(cbind(txt_1col, read.table(text=txt_2col_End[-33])))

colnames <- txt1[1] %>% 
  str_squish() %>% 
  str_split(" ") %>% 
  unlist

colnames(df)[-1] <- colnames
colnames(df)[1] <-  "Car Model"


  

  




```





## gene expression in different cancer cell lines

```{r}
library(data.table)
library(tidyverse)

# fread can directly recognize .gz file
file <- "./data/CCLE_RNAseq_rsem_transcripts_tpm_20180929.txt.gz"
CCLE_DATA <- fread(file=file, data.table = F)

index <- grep("BREAST", names(CCLE_DATA)) # Return the col number of matched items.

data <- CCLE_DATA %>% 
    select(c(2, index)) %>% 
    separate(transcript_id,into = "transcript_id",sep = "\\.") %>% 
    column_to_rownames("transcript_id") %>% 
    t() %>%                               # transpose, the result is matrix
    data.frame() %>%                      # convert to data.frame
    rownames_to_column("cellline") %>%    
    separate(cellline, into="cell_line", sep = "\\_")

data_anno <- fread("./data/Cell_lines_annotations_20181226.txt")
names(data_anno)[1]  <- "CCLE_name"
names(data_anno)[2] <-  "Cell_line_primary_name"



breast_ccle_index <- data_anno %>%  
  filter(Site_Primary=="breast") %>% 
  select(CCLE_name,Name) %>%
  separate(CCLE_name,into = "celline", sep = "\\_") %>% 
  column_to_rownames("celline")

# More like the named vector. Vector[name] = value

data <- cbind(cell_name=breast_ccle_index[data$cell_line, 1],data)



save(data, file = "./output/CCLE_BREAST_PLOT.Rdata")


gene_id <- "ENST00000373020"
test <- data[,c("cell_name",gene_id)]
names(test)[2] <- "gene_id"

# fct_reorder: reorder column "cell_name" based the gene_expression of the given gene_id
p <-  ggplot(data = test,aes(x=forcats::fct_reorder(cell_name,desc(gene_id)),y=gene_id))+
  geom_bar(stat='identity', aes(fill=cell_name))+
  theme_bw()+
  ylab(paste0(gene_id," (TPM)"))+
  xlab(NULL)+
  theme(legend.position = "none")+
  theme(axis.text.x = element_text(angle = 45, hjust = 1))






```

## Connecting across missing values with geom_line
- https://stackoverflow.com/questions/9617629/connecting-across-missing-values-with-geom-line


```{r}
library(tidyverse)

dfr <- data.frame(
  x = 1:10,
  y = runif(10)
)

dfr[c(3, 6, 7), "y"] <- NA

ggplot(dfr, aes(x,y)) + 
  geom_point() +
  geom_line(data=dfr[!is.na(dfr$y),])

ggplot(dfr[!is.na(dfr$y),], aes(x,y)) + 
  geom_point() +
  geom_line()




```
## y叔：replace backgroud color of a photho

```{r}
library(magick)
library(cowplot)
library(ggplotify)

file="Tommi.JPG"
x= image_read(file)
image_fill(x, "white", fuzz=20) -> y
z <- cowplot::plot_grid(as.ggplot(x), as.ggplot(y), ncol=2)
window()
grid.txt("太帅了， 不给你们看"， gp=gpar(cex=4))
         





```






## Cont maxtrix input
- featureCounts -> DESeq2
- deseq2-anslysis template 
  https://gist.github.com/stephenturner/f60c1934405c127f09a6
  
  df <- data.frame(x = c(NA, "a?b", "a.d", "b:c"))
  df %>% separate(x, c("A","B"), sep = "([\\.\\?\\:])")

```{r}
library(tidyverse)
library(data.table)
library(DESeq2)

if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("rnaseqGene")
BiocManager::install(version = "3.9")
browseVignettes("rnaseqGene")

# Import and pre-process
# data from featureCounts -T 12  -t exon -g gene_id -a genes.gtf -o counts.txt  GSM*.bam
file="./data/SRR4820707_counts_all.txt"

countdata <- read.table(file=file, header = T, row.names=1)
countdat[1:10, 1:8]
countdata1 <- countdata

countdata <- countdata %>% 
  select_all(~gsub("_.*\\.bam$", "", .))  %>% 
  select(6:ncol(.)) %>% 
  as.matrix()

  
# Creat the coldata
coldata <- read.table("./data/SRR4820707_sampleInfo.txt", header = F) %>% 
  as_tibble() %>% 
  column_to_rownames("V1") %>% 
 dplyr::rename("condition" = "V2", "replicate" = "V3")

# Check if the colnames of countdat match rownames of coldata
colnames(countdata) == rownames(coldata)

# Create the DESeqDataSet
dds <- DESeqDataSetFromMatrix(countData = countdata, colData = coldata, design = ~replicate+condition)

# Pre-filtering  I added @7/15/19
keep <- rowSums(counts(dds) > 8)
dds <- dds[keep, ]

# Note on factor levels (I added @7/15/19)

#dds$condition <- factor(dds$condition, levels=c("untreated", "treated")) # factor function
dds$condition <- relevel(dds$condition, ref="shLUC")

dds <- estimateDispersionsGeneEst(dds)
# Run the DESeq pipeline
dds <- DESeq(dds)


# Plot dispersions
png("qc-dispersions.png", 1000, 1000, pointsize=20)
plotDispEsts(dds, main="Dispersion plot")
dev.off()

# Regularized log transformation for clustering/heatmaps, etc
rld <- rlogTransformation(dds)
head(assay(rld))
hist(assay(rld))

res <- results(dds)



```

## gene ID Convert 

```{r}
library(biomaRt)

# The target genes for conversion
gene_ids <- c("TP53", "SMAD4","BRCA1")

# Select a database
mart=useMart(biomart="ensembl", dataset="hsapiens_gene_ensembl")
ensembl=useMart(biomart = "ensembl", dataset = "hsapiens_gene_ensembl")



```
## Going from a human readable Excel file to a machine-readable csv with {tidyxl}

- https://www.brodrigues.co/blog/2018-09-11-human_to_machine/


```{r}
library(tidyverse)
library(tidyxl)

file <- c("./data/time_use.xlsx")
time_use_xl <-  xlsx_cells(file)
class(time_use_xl)
head(time_use_xl)


sheets <-  xlsx_sheet_names(file) %>% 
    keep(grepl(pattern = ".*day$", .))  # grep() will return the index in the sheet vector, which is not applicable in the pipe.

extract_data <- function(sheet){
    activities <- sheet %>%
        filter(col == 2) %>%
        select(row, character) %>%
        filter(row %in% seq(6,58)) %>%
        dplyr::rename(activities = character) %>%
        select(-row)
    
    cols_to_extract <-  sheet %>% 
      filter(grepl("Population who complete.", character)) %>% 
      pull(col)
    
    headers_pos <-  cols_to_extract - 1
    
    headers <- sheet %>% 
      filter(col %in% headers_pos, row == 3) %>% 
      pull(character)
    
    cols_to_extract %>% 
      map(~filter(sheet, col %in% .)) %>% 
      map(~select(., sheet, address, row, col, character)) %>% 
      map(~filter(., row %in% seq(6,58))) %>% 
      map(~select(., character)) %>% 
      map(.x = ., .y =headers, ~mutate(.x, "population" = .y)) %>% 
      map(., ~bind_cols(acitivities, .)) %>% 
      bind_rows()
    
}


clean_data <- sheets %>%
    map(~filter(time_use_xl, sheet %in% .)) %>%  # sheet is the name of the first column of time_use_xl
    set_names(sheets) %>% 
    map(extract_data) %>% 
    map2(.x=., .y= sheets, ~mutate(.x, "day" = .y)) %>% 
    bind_rows() %>% 
    select(day, population, activities, time = character)

glimpse(clean_data)

#--- test data step by step---------------------------------------------------------------
mysheet <- clean_data[[1]]

activities <- mysheet %>%
        filter(col == 2) %>%                        # select the 2nd column
        select(row, character) %>%
        filter(row %in% seq(6,58)) %>%              # select row 6-58
        dplyr::rename(activities = character) %>%   # without dplyr:: will cause error.
        select(-row) # drop the column named row

  # I wrote  Population  as Populaition, I figure this out for 2 hours by replacing one word a time.
cols_to_extract <- mysheet %>% 
  filter(grepl("Population who completed.*", character)) %>% 
   pull(col)    # pull : cols_to_extract is an integer vector, selet(col) will be a tbl_df

headers_pos <-  cols_to_extract - 1


headers <- mysheet %>% 
      filter(col %in% headers_pos, row == 3) %>% 
      pull(character)

 t <- cols_to_extract %>% 
      map(~filter(mysheet, col %in% .)) %>% 
      map(~select(., sheet, address, row, col, character)) %>% 
      map(~filter(., row %in% seq(6,58))) %>% 
      map(~select(., character)) %>% 
      map(.x = ., .y =headers, ~mutate(.x, "population" = .y)) %>% 
      map(., ~bind_cols(acitivities, .)) %>% 
      bind_rows()
    
s <-  mysheet %>% 
    filter(col %in% cols_to_extract)

---------------------------------------------------------------------------------------------

```



## Using corrr
- https://cran.r-project.org/web/packages/corrr/vignettes/using-corrr.html

ls("package:corrr")
 [1] "as_cordf"     "as_matrix"    "correlate"    "dice"         "fashion"     
 [6] "first_col"    "focus"        "focus_"       "focus_if"     "network_plot"
[11] "pair_n"       "rearrange"    "retract"      "rplot"        "shave"       
[16] "stretch"    

- EASY CORRELATION MATRIX ANALYSIS IN R USING CORRR PACKAGE https://www.datanovia.com/en/blog/easy-correlation-matrix-analysis-in-r-using-corrr-package/
-

```{r}
library(corrr)


d <- correlate(mtcars, quiet = T)
class(d)
d

library(dplyr)
d %>%
  filter(cyl > .7) %>% 
  select(rowname, mpg, cyl, disp)

library(purrr)
d %>% 
  select(-rowname) %>% 
  map_dbl(~ mean(., na.rm = TRUE))

# Innternal changes (cor_df out):
# 
#  shave() the upper or lower triangle (set to NA).
#  rearrange() the columns and rows based on correlation strengths.
#  Reshape structure (tbl or cor_df out):
# 
# focus() on select columns and rows.
#   stretch() into a long format.
#   Output/visualizations (console/plot out):
# 
# fashion() the correlations for pretty printing.
#   rplot() a shape for each correlation.
#   network_plot() a point for each variable, joined by paths for correlations.

d %>%
  focus(mpg:drat, mirror = T) %>%  # Focus only on mpg:drat
  shave() %>% # Remove the upper triangle
  fashion()   # Print in nice format 

d %>%
  focus(mpg:drat, mirror = T) %>%  # Focus only on mpg:drat
  shave() %>% # Remove the upper triangle
  rplot   # plot 


d %>%
  focus(mpg:drat, mirror = TRUE) %>%
  rearrange(absolute = FALSE) %>% 
  shave() %>%
  rplot()

d %>%
  focus(mpg:drat, mirror = TRUE) %>%
  rearrange(absolute = FALSE) %>% 
  network_plot() # keep shave() lkie rplot will cause error.

# https://github.com/tidymodels/corrr/issues/29----------------------------------
# In the meantime, you have some nice options in other packages (just mentioning for anyone else who comes across this):
# 
# RcmdrMisc::rcorr.adjust()
# psych::corr.test()
# After using these, you can make use of the new function, as_cordf() (available in dev version only). If you have a matrix, even of p-values, use as_cordf() to transform it to corrr's correlation data frame format, which you can then feed into any other corrr functions like rplot().
# ---


```
## PCA tutorial from DataCamp
 - https://www.datacamp.com/community/tutorials/pca-analysis-r
 
```{r}
mtcar.pca <- prcomp(mtcars[, c(1:7, 10, 11)], center = TRUE, scale. = T)
summary(mtcar.pca)

library(devtools)
install_github("vqv/ggbiplot")
library(ggbiplot)

ggbiplot(mtcar.pca)


mtcars.country <- c(rep("Japan", 3), rep("US",4), rep("Europe", 7),rep("US",3), "Europe", rep("Japan", 3), rep("US",4), rep("Europe", 3), "US", rep("Europe", 3))


ggbiplot(mtcar.pca, ellipse = T, obs.scale = 1, var.scale = 1, var.axes = F, 
        labels = rownames(mtcars), groups=mtcars.country )















```
 

Outputting Plots with map Functions
## Outputting Plots with map Functions walk

```{r}
library(tidyverse)

plot_points <- function(t) ggplot(t, aes(mpg, wt)) + geom_point()

mtcars %>%                 # motortrend car dataset
  group_by(cyl) %>%       # group by cylinder
  nest(.key = "data") %>% # nest by cylinder
  mutate(plot = map(data, plot_points)) %>% # we made a graph for each subset
  pull(plot) %>% #pull the plots out as a list
  walk(., print) # print them

```
## Save Plots from walk with walk2/pwalk

```{r}
# make a simple plotting function
plot_points <- function(t) ggplot(t, aes(mpg, wt)) + geom_point()

mtc_plot_n_save <-
  mtcars %>%                 # motortrend car dataset
    group_by(cyl) %>%       # group by cylinder
    nest(.key = "data") %>% # nest by cylinder
    mutate(plot = map(data, plot_points)) %>% # we made a graph for each subset
    mutate(filepath = map_chr(cyl, ~ file.path("figures", paste0(.x, ".png")))) #make the save paths


walk2(mtc_plot_n_save$plot, mtc_plot_n_save$filepath, ~ ggsave(.y, .x)) # save them out



```

```{r}
# make a simple plotting function
plot_points <- function(t) ggplot(t, aes(mpg, wt)) + geom_point()

#function that prints cylinder info, ggplot, and table 
print_func <-
  function(info, plot, table, filepath) {
    print(paste0("This is for ", info, " cylinders./n"))
    print(plot)
    ggsave(filepath)
    print(table)
    
  }

mtc_plot_n_save <-
  mtcars %>%                 # motortrend car dataset
    group_by(cyl) %>%       # group by cylinder
    nest(.key = "data") %>% # nest by cylinder
    mutate(plot = map(data, plot_points)) %>% # we made a graph for each subset
    mutate(filepath = map_chr(cyl, ~ file.path("figures", paste0(.x, ".png")))) # make the save paths

# print all the info
pwalk(list(info = mtc_plot_n_save$cyl, plot = mtc_plot_n_save$plot, table = mtc_plot_n_save$data, filepath = mtc_plot_n_save$filepath), print_func) 



```


## Import a Directory of CSV Files at Once Using {purrr} and {readr}




